### PR TITLE
Add support for the v1 alerts api.

### PR DESCRIPTION
**Pull request by hhellyer**
This PR adds support for the alerts api.

- Adds code to call the new alerts api.
- Adds testcases.
- Adds example.

### CODE REVIEW

File: rest/model/alerting/alert.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package alerting
```
Comment (apantelopoulosNS1) [inline rest/model/alerting/alert.go:1]: oh nice, more models! its not like that we did not have enough already :D

Comment (apantelopoulosNS1) [reply rest/model/alerting/alert.go:1]: not a jab at the PR but rather than the fact that we have a 1000 models!

File: rest/_examples/alerts.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	api "gopkg.in/ns1/ns1-go.v2/rest"
+	"gopkg.in/ns1/ns1-go.v2/rest/model/alerting"
+	"gopkg.in/ns1/ns1-go.v2/rest/model/dns"
+	"gopkg.in/ns1/ns1-go.v2/rest/model/monitor"
+)
+
+var client *api.Client
+
+// Helper that initializes rest api client from environment variable.
+func init() {
```
Comment (apantelopoulosNS1) [inline rest/_examples/alerts.go:20]: Is this called in the example somewhere?

**Review by apantelopoulosNS1 — DISMISSED** [review]

File: rest/_examples/alerts.go (hunk 2 above)
Comment (hhellyer) [reply rest/_examples/alerts.go:20]: If I'm 100% honest this is here because I started by copying one of the other examples and wanted it to follow our existing pattern.
It will always be called as go ensures the `init()` function for a file is always run:
https://go.dev/doc/effective_go#init
(The internet seems rather divided on whether it's a bad thing to use which in general but it's probably ok for samples like this.)

File: rest/_examples/zones.go [hunk 3]
```diff
@@ -62,7 +62,7 @@ func main() {
 	}
 
 	// Add an A record with a single static answer.
-	orchidRec := dns.NewRecord(domain, "orchid", "A")
+	orchidRec := dns.NewRecord(domain, "orchid", "A", nil, nil)
```
Comment (shane-ns1) [inline rest/_examples/zones.go:65]: Is `dns.NewRecord()` a call that is part of the Go API? If so, extending it to require additional parameters is a breaking change, meaning we'd have to go from version 2.12.2 to version 3.

If we don't want version 3 (which probably we do not?) and this is part of our Go API, then we should consider adding a new function instead. `dns.NewRecordEx()` or something like that, if we follow Windows-style naming conventions, or `dns.NewRecord2()` if we follow POSIX-style naming convention (I prefer the Windows apporach in this case).

File: rest/model/alerting/alert.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	ZoneNames       []string        `json:"zone_names"`
+}
+
+var (
+	zoneAlertType   string = "zone"
+	recordAlertType string = "record"
+)
+
+// TODO - Allow zones/notifier lists to be passed as structs?
+//   - Multiple constructors?
```
Comment (shane-ns1) [inline rest/model/alerting/alert.go]: I'm not a fan of TODO comments in our code. If there is a TODO file for our Go API, then I guess we can put it there. Otherwise we can make a GitHub issue with this idea.

File: rest/client.go [hunk 5]
```diff
@@ -13,9 +13,9 @@ import (
 )
 
 const (
-	clientVersion = "2.12.2"
+	clientVersion = "2.13.0"
 
-	defaultBase                   = "http://localhost:80"
+	defaultBase                   = "https://api.nsone.net"
```
Comment (apantelopoulosNS1) [inline rest/client.go:18]: why did you change this?

**Review by apantelopoulosNS1 — DISMISSED** [review]

File: rest/alert.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	// Update the alerts fields with data from api (ensure consistent)
+	resp, err := s.client.Do(req, &alert)
+	if err != nil {
+		return resp, err
+	}
+
+	return resp, nil
+}
+
+// Delete immediately deletes an existing alert.
```
Comment (shane-ns1) [inline rest/alert.go]: The word "immediately" might make my Spidey-Sense tingle if I was a customer reading this. Why do we say "immediately"? Is there a _non_-immediate way to delete an alert? Do other commands not work immediately? :laughing:

File: rest/alert_test.go [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"gopkg.in/ns1/ns1-go.v2/mockns1"
+	"gopkg.in/ns1/ns1-go.v2/rest/model/alerting"
+
+	api "gopkg.in/ns1/ns1-go.v2/rest"
+)
+
+func strPtr(str string) *string {
```
Comment (shane-ns1) [inline rest/alert_test.go:15]: Annoying that such a function is needed. I should add this to my "Why I Hate Go" presentation!

File: rest/alert_test.go [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+			// Pass by value here. alertToCreate will be modified by client.Alerts.Create
+			require.Nil(t, mock.AddAlertCreateTestCase(nil, nil, alertToCreate, *alertResponse))
+
+			_, err := client.Alerts.Create(&alertToCreate)
+			require.Nil(t, err)
+			// alerttoCreate will be modified by the call to Create and fully populated.
+			compareAlerts(t, alertResponse, &alertToCreate)
+		})
+
+		// TODO Add test for error response when updating fields that can't be changed.
```
Comment (shane-ns1) [inline rest/alert_test.go]: Is this difficult? Why isn't it in the PR?

**Review by shane-ns1 — CHANGES_REQUESTED** [review]
I have a few questions, nothing very critical.

I am concerned about possibly changing the API without bumping the major version number, but I maybe this concern is misplaced.

File: rest/_examples/zones.go (hunk 3 above)
Comment (hhellyer) [reply rest/_examples/zones.go:65]: Ah - I should have put a note in the PR.
I only changed this because dns.NewRecord has taken 5 params since https://github.com/ns1/ns1-go/pull/220 some time ago and the example was already broken.
I probably should have raised a separate issue. I can move this to a different issue if you like.

File: rest/alert.go (hunk 6 above)
Comment (hhellyer) [reply rest/alert.go]: I copied that from the notifier list sdk code - I was trying to blend in. Removed.

File: rest/alert_test.go (hunk 8 above)
Comment (hhellyer) [reply rest/alert_test.go]: Removed, it's not difficult it's also not an issue on alert creation - I copied this from either update or replace which does test this.

File: rest/_examples/zones.go (hunk 3 above)
Comment (shane-ns1) [reply rest/_examples/zones.go:65]: No worries. We should have made a major version bump or added a new "constructor" at that time, but we are where we are. :laughing:

Since this is just an example fix from your PR, I think we're good.

**Review by shane-ns1 — APPROVED** [review]
Looks good. Thanks for the attention to detail. :+1:
